common/src/main/java/org/ergoplatform/appkit/Eip29Attachment.java

```diff
@@ -0,0 +1,221 @@
+    }
+
+    /**
+     * @return array R4-R9 to use with OutboxBuilder#registers
+     */
+    public ErgoValue<?>[] getOutboxRegistersForAttachment() {
+        ErgoValue<?>[] registers = new ErgoValue[6];
+
+        for (int i = 0; i < registers.length - 1; i++) {
+            registers[i] = ErgoValue.unit();
+        }
+        registers[5] = getErgoValue();
+        return registers;
+    }
+
+    /**
+     * @param r9 Ergo value to create the attachment object from, usually stored in register 9 of boxes
+     * @return object representing the attachment
+     */
+    public static Eip29Attachment createFromErgoValue(ErgoValue<?> r9) {
+        String illegalArgumentException = "R9 must be of pair (Coll[0x50, 0x52, 0x50], Tuple2(Int, Coll[Byte]), actual: ";
+
+        if (!(r9.getValue() instanceof Tuple2)) {
+            throw new IllegalArgumentException(illegalArgumentException + r9.getValue().getClass().getSimpleName());
+        }
+
+        Tuple2 attachmentWrapper = (Tuple2) r9.getValue();
+        if (!(attachmentWrapper._1 instanceof Coll) || !(attachmentWrapper._2 instanceof Tuple2)) {
+            throw new IllegalArgumentException(illegalArgumentException + r9.toHex());
+        }
+
+        byte[] magicBytes = JavaHelpers$.MODULE$.collToByteArray((Coll<Object>) attachmentWrapper._1);
+        if (!Arrays.equals(Eip29Attachment.MAGIC_BYTES, magicBytes)) {
+            throw new IllegalArgumentException(illegalArgumentException + r9.toHex());
+        }
+
+        Tuple2 attachmentValue = (Tuple2) attachmentWrapper._2;
+        if (!(attachmentValue._1 instanceof Integer) || !(attachmentValue._2 instanceof Coll)) {
+            throw new IllegalArgumentException(illegalArgumentException + r9.toHex());
+        }
+
+        return createFromAttachmentTuple(attachmentValue);
+    }
+
+    private static Eip29Attachment createFromAttachmentTuple(Tuple2 attachmentTuple) {
+        int typeConstant = (int) attachmentTuple._1;
+        Eip29Attachment.Type attachmentType = Eip29Attachment.Type.fromTypeRawValue(typeConstant);
+        byte[] attachmentContent = JavaHelpers$.MODULE$.collToByteArray((Coll<Object>) attachmentTuple._2);
+
+        switch (attachmentType) {
+            case PLAIN_TEXT:
+                return new PlainTextAttachment(attachmentContent);
+            case MULTI_ATTACHMENT:
+                return new MultiAttachment(attachmentContent);
+            default:
+                return new Eip29Attachment(typeConstant, attachmentContent);
+        }
+    }
+
+    /**
+     * Type of attachment
+     */
+    public enum Type {
+        /**
+         * This attachment is a list of attachments
+         */
+        MULTI_ATTACHMENT,
+        /**
+         * This attachment is a plain text, see {@link PlainTextAttachment#getText()}
+         */
+        PLAIN_TEXT,
+        /**
+         * Attachment type is undefined or unknown
+         */
+        UNDEFINED;
```

Comment:
Enum type can better be checked by compilers if switch/when statements are exhaustive, so it is always better to use an enum type instead of integer constants in my opinion. The raw integer constant should be still provided in case clients need to persist the original data (this can be done by using Eip29Attachment.getTypeRawValue).

Reply:
Ok, then I think we need to draw a distinction between Undefined and Unknown. Undefined is a kind of null value, which is not assigned. Unknown is a first value which cannot be interpreted by the current version. And it can be used in conditions like `if (tpe >= Unknown) ...` 
What do you think?